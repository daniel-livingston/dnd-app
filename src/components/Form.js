import React from "react";
import PropTypes from "prop-types";
import FormItem from "./FormItem";
import FormButton from "./FormButton";

const propTypes = {
	header: PropTypes.string,
	items: PropTypes.arrayOf(
		PropTypes.exact({
			title: PropTypes.string.isRequired,
			description: PropTypes.string,
			inputType: PropTypes.string.isRequired,
			inputValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
			inputOnChange: PropTypes.func.isRequired,
			inputPlaceholder: PropTypes.string.isRequired,
			selectOptions: PropTypes.arrayOf(
				PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired
			),
		})
	).isRequired,
	button: PropTypes.exact({
		onClick: PropTypes.func.isRequired,
		id: PropTypes.string,
		name: PropTypes.string.isRequired,
		justify: PropTypes.string,
		width: PropTypes.string,
	}),
};

const Form = (props) => (
	<form className='Form' onSubmit={(e) => e.preventDefault()}>
		{props.header && <h2 className='Form-header'>{props.header}</h2>}
		{props.items.map((item) => (
			<FormItem {...item} key={item.title} />
		))}
		{props.button && <FormButton {...props.button} />}
	</form>
);

Form.propTypes = propTypes;

export default Form;
