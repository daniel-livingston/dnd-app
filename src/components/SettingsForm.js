import React from "react";
import { connect } from "react-redux";
import Form from "./Form";
import { setTheme } from "../actions/theme";
import { THEMES } from "../constants";

export class SettingsForm extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			testing: true,
		};
	}
	onThemeChange = (e) => {
		this.props.setTheme(e.target.value);
	};
	render() {
		return (
			<div className='SettingsForm'>
				<Form
					header='Appearance'
					items={[
						{
							title: "Theme",
							description: "Choose the theme for the app.",
							inputType: "select",
							inputValue: this.props.theme.name,
							inputOnChange: this.onThemeChange,
							inputPlaceholder: "Theme",
							selectOptions: THEMES.map((theme) => theme.name),
						},
					]}
				/>
			</div>
		);
	}
}

const mapStateToProps = (state) => ({
	theme: state.theme,
});

const mapDispatchToProps = (dispatch) => ({
	setTheme: (theme) => dispatch(setTheme(theme)),
});

export default connect(mapStateToProps, mapDispatchToProps)(SettingsForm);
