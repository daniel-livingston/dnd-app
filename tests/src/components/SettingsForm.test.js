import React from "react";
import { mount, shallow } from "enzyme";
import { Provider } from "react-redux";
import Form from "../../../src/components/Form";
import SettingsForm, {
	SettingsForm as DisconnectedSettingsForm,
} from "../../../src/components/SettingsForm";
import store from "../../../src/store/configureStore";

const props = {
	theme: {
		name: "Test",
	},
	setTheme: jest.fn(),
};

describe("<SettingsForm>", () => {
	test("renders correctly", () => {
		const wrapper = shallow(<DisconnectedSettingsForm {...props} />);
		expect(wrapper).toMatchSnapshot();
	});

	test("changes theme on setting change", () => {
		const wrapper = mount(
			<Provider store={store}>
				<SettingsForm />
			</Provider>
		);

		const onThemeChange = wrapper.find(Form).props().items[0].inputOnChange;
		const values = wrapper.find(Form).props().items[0].selectOptions;

		onThemeChange({ target: { value: values[1] } });
		expect(store.getState().theme.name).toEqual(values[1]);

		onThemeChange({ target: { value: values[0] } });
		expect(store.getState().theme.name).toEqual(values[0]);
	});
});
