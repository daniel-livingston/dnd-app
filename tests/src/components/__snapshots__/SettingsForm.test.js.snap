// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsForm> renders correctly 1`] = `
<div
  className="SettingsForm"
>
  <Form
    header="Appearance"
    items={
      Array [
        Object {
          "description": "Choose the theme for the app.",
          "inputOnChange": [Function],
          "inputPlaceholder": "Theme",
          "inputType": "select",
          "inputValue": "Test",
          "selectOptions": Array [
            "Light",
            "Dark",
          ],
          "title": "Theme",
        },
      ]
    }
  />
</div>
`;
